### PR TITLE
[FLINK-2136] Adding DataStream tests for Scala API

### PR DESCRIPTION
**Pull request by gaborhermann**
* Added tests for scala DataStream
* Added tests for setting parallelism
* Added tests for operators
* Fixed some small bugs in scala API
* Added exception throwing to streaming functions

### CODE REVIEW

**Comment by mbalassi** [comment]
Thanks, this is definitely a must have.

**Comment by mbalassi** [comment]
Please fix the scala style issues. :)

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java [hunk 1]
```diff
@@ -617,7 +617,7 @@ protected ExecutionConfig getExecutionConfig() {
 	 */
 	public <R> SingleOutputStreamOperator<R, ?> fold(R initialValue, FoldFunction<OUT, R> folder) {
 		TypeInformation<R> outType = TypeExtractor.getFoldReturnTypes(clean(folder), getType(),
-				Utils.getCallLocationName(), false);
+				Utils.getCallLocationName(), true);
```
Comment (gaborhermann) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java]: I set allowMissing to true as it was set at flatMap. Leaving it on false caused the call fold in the Scala DataStream fail.

File: flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/operators/AbstractUdfStreamOperator.java [hunk 2]
```diff
@@ -84,6 +84,10 @@ public void confirmCheckpointCompleted(long checkpointId, long timestamp) throws
 		}
 	}
 
+	public F getUserFunction() {
```
Comment (gaborhermann) [inline flink-staging/flink-streaming/flink-streaming-core/src/main/java/org/apache/flink/streaming/api/operators/AbstractUdfStreamOperator.java]: Added a public getter, thus the StreamGraph creation from APIs is more testable.

**Comment by gaborhermann** [comment]
Fixed the style issues and added more tests.

**Comment by StephanEwen** [comment]
General questions: Are type hints implemented already in the streaming API?

**Comment by mbalassi** [comment]
@StephanEwen: The `SingleOutpustStreamOperator` has the returns method, similar to the batch API.

File: flink-staging/flink-streaming/flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    assert("testReduce" == dataStream2.getName)
+
+    val connected = dataStream1.connect(dataStream2)
+      .flatMap(
+    { (in, out: Collector[Long]) => }, { (in, out: Collector[Long]) => }
+    ).name("testCoFlatMap")
+    assert("testCoFlatMap" == connected.getName)
+
+    val fu: ((Long, Long) => Long) =
+      (x: Long, y: Long) => 0L
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala]: Let us rename this to `func`. :)

File: flink-staging/flink-streaming/flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+
+    val connected = src1.connect(src2)
+
+    val group1 = src1.groupBy(0)
+    val group2 = src1.groupBy(1, 0)
+    val group3 = src1.groupBy("_1")
+    val group4 = src1.groupBy(x => x._1)
+
+    assert(isPartitioned(graph.getStreamEdge(group1.getId, createDownStreamId(group1))));
+    assert(isPartitioned(graph.getStreamEdge(group2.getId, createDownStreamId(group2))));
```
Comment (mbalassi) [inline flink-staging/flink-streaming/flink-streaming-scala/src/test/scala/org/apache/flink/streaming/api/scala/DataStreamTest.scala]: Trivial: semicolon.

**Comment by mbalassi** [comment]
Let me merge this for the release now.

**Comment by mbalassi** [comment]
The iteration test sometimes fails on travis, so  removed that.
https://travis-ci.org/mbalassi/flink/jobs/65594738

**Comment by gaborhermann** [comment]
Thanks for merging. I removed the line that failed the test (I mistakenly checked a non-deterministic id).
https://github.com/gaborhermann/flink/commits/FLINK-2136

**Comment by mbalassi** [comment]
Please open a PR with it. Could you give FLINK-2180 a try in the meantime please? It looks similar on paper. You can add then to the same pull request of course. :)

**Comment by gaborhermann** [comment]
Of course. I took up the JIRA issue.

**Pull request closed without merging**